### DISCUSSION TITLE
Setting a LGCP model with a 'strange' domain

### DISCUSSION THREAD

**Discussion by osupplisson** [inlabru-org/inlabru#253] [Q&A] [answered]
Dear inlabru users,

I am having some issues in fitting a LGCP model and maybe some of you might be able to give me some advice.

I am trying to model the spatial location of some hospital admissions happening in a region (which also happens to be a single city), spanning around 2 ,095 km². I am observing around 7,000 admissions and I know that I am missing some of them which led me to specify a thinned LGCP.

In the figure below, I am showing the spatial domain along with the subdomain (in black) in which it is possible to observe a case:

![image](https://github.com/user-attachments/assets/f1ec3437-db68-455a-b8e4-0fcf5852b1a0)

The subdomain is a set of linestring produced by several intersection of rasters and the roadnetwork from open street map:

![image](https://github.com/user-attachments/assets/adc29a4d-30c8-4a66-aa8b-fd3dcf9fe965)

Because the area is large, I am using a quite coarse mesh, resulting in 1,903 nodes:

``
mesh_space <- fm_mesh_2d_inla(
  loc.domain = sampling_domain,
  crs = fmesher::fm_crs(kmproj),
  offset = c(1, 4),
  max.edge = c(1, 4),
  cutoff = 1,
  min.angle = 21,
  boundary = boundary_hcmc
)
``

![image](https://github.com/user-attachments/assets/a807ed44-6050-4ea5-baa5-e6f520cb39fc)

My main concern is that cases can only be observed in very specific locations while (I believe) the model should be defined for the whole domain. To circumvent this issue, I tried to find a way to define the intensity such that it tends towards 0 for areas in which we should not observe any case and to something else >0 for admissible areas. To that end, I defined the components and formula in the code chunk below. 

```
# Components
cmp_baseline <- ~ 0 +
#intercept common to all location
  baselineintercept(1,
                    model = "linear",
                    mean.linear = 0,
                    prec.linear = 0.1)  +
#intercept only for admissible areas: 1_{living area(s)=1}*beta0
  intercept(
    eval_spatial(living_area, where = geometry, layer = 1),
    model = "linear",
    mean.linear = 0,
    prec.linear = 0.1
  )  +
#Explanatory variable with time-specific parameter: 1_{living area(s)=1}*x(s,t)*beta_{t}
  flood(
    idtime,
    weights = eval_spatial(living_area, where = geometry, layer = 1) *
      eval_spatial(
        data = flood,
        where = geometry,
        layer = idtime
      ),
    model = "rw2",
    scale.model = TRUE,
    constr = FALSE,
    hyper = list(prec = list(
      prior = 'pc.prec', param = c(u = 1, a = 0.01)
    ))
  ) +
#Spatial GRF:  1_{living area(s)=1}*f(s)
  space(
    geometry,
    weights = eval_spatial(living_area, where = geometry, layer = 1),
    model = spacespde
  ) +
#Temporal GMRF: 1_{living area(s)=1}*f(t)
  time(
    idtime,
    weights = eval_spatial(living_area, where = geometry, layer = 1),
    model = "rw2",
    scale.model = TRUE,
    constr = TRUE,
    hyper = list(prec = list(
      prior = 'pc.prec', param = c(u = 1, a = 0.01)
    ))
  ) +
#Components for the thinning process
  sigma(
    1,
    mean.linear = 0,
    prec.linear = 1,
    marginal = bru_mapper_marginal(qexp, pexp, dexp, rate = 80)
  ) +
  xi(
    1,
    mean.linear = 0,
    prec.linear = 1,
    marginal = bru_mapper_marginal(qgamma, pgamma, dgamma, shape = 1, rate = 1)
  )

##For admissible area, I expect the intercept to 'offset' the global intercept, such that the intensity becomes > 0 for admissible area while being ~0 for areas in which we cannot observe cases.
```

The definition of the components rely on a raster (`living_area`) which takes values 1 for admissible area and 0 otherwise. To avoid numerical issue, I made a 10 meters buffer around the linestring to build this raster

![image](https://github.com/user-attachments/assets/9d13df26-71d3-4ceb-b6a0-cfcdc365cc73)

The formula and likelihood are defined as follow:
```
# Formula--------------------------------------------------------------
fml <- geometry + idtime ~   baselineintercept +
#thinning process
  logdetprob(geometry, sigma, xi) +
#normalising offset: equal to 0 outside admissible areas and log(Pop) otherwise
  logpop(geometry) +
  intercept +
  flood +
  space +
  time

domain_subdivided <- fmesher::fm_subdivide(mesh_space, n = 10)

# Likelihood --------------------------------------------------------------
likelihood_baseline <-  like(
  data =  df_final,
  family = "cp",
  formula = fml,
  domain = list(geometry = domain_subdivided , idtime = set_time),
  sampler = list(geometry = boundary_hcmc, idtime = set_time)
)
```
With the current model definition, I am running into numerical problems similar to what is described here: https://github.com/inlabru-org/inlabru/discussions/240. I believe this comes from a threefold sources: the way I defined the model + the coarse mesh + the high resolution of the covariate (25m for flood and 10m for population). 

Using `fmesher::fm_subdivide()` function, with `n` set to 6, I was able to run the model for a binary explanatory variable and without the log(population) offset. With the log(population) offset and the covariate defined as a proportion, it seems I have to set it to more than 8  (I am currently testing for n=10 and n=12).  Given that I also have a temporal dimension (12 periods), the model is starting to be really long to set (especially the spatial GRF component takes hours at the 'Evaluate predictor linearisation' step).

Any idea to handle this better would be super helpful !

As a sidenote: I also gave it a try with the MetricGraph package (https://davidbolin.github.io/MetricGraph/). I also ended up with a very long runtime just to set the mesh (I killed the HPC jobs after 2 days). Besides, the package is not yet ready for inlabru, which I need to include the thinning process.

Best, 
Olivier

**Comment by finnlindgren** [comment]
The MetricGraph devel version which is about to be submitted to CRAN is reportedly much faster for some operations, and it will have more direct inlabru support (there is also further inlabru support work in progress that should be in the next version after the one about to be submitted).

I notice one issue in your code comments:
```
#normalising offset: equal to 0 outside admissible areas and log(Pop) otherwise
```
To get a "not possible" effect you would need `-Inf`, not `0`, as it's applied on the log-intensity; a 0 additive offset means "intensity multiplier = 1", which is likely not at all what you want or need, and is likely to be a cause for numerical issues (using -Inf would also be problematic for this reason).
The proper way to handle known inadmissible regions is to specify a sampler polygon/s that only covers admissible regions; you'd need the intersection polygon of your current sampler boundary and the polygons for admissible regions.

What inlabru and fmesher versions did you try? Just the cran versions? I'm nearly about to submit a new inlabru version to cran, that may have some speedups (also, some function naming changes; `like()` will still be around for a while, but `bru_obs()` is the new "canonical" name for it.

**Comment by osupplisson** [comment]
Hi Finn,

Thank you very much for your reply.

Regarding inlabru/fmesher, I am using a quite recent devel version: `inlabru_2.11.1.9022` `fmesher_0.2.0`. For the MetricGraph, I indeed noticed the recent upgrade for the mesh building function in the devel version but have not tried it yet.

> To get a "not possible" effect you would need -Inf, not 0, as it's applied on the log-intensity; a 0 additive offset means "intensity multiplier = 1", which is likely not at all what you want or need, and is likely to be a cause for numerical issues (using -Inf would also be problematic for this reason).

I used "intensity multiplier = 1" for inadmissible regions on purpose (in that case intensity = baselineintercept ), with the idea to let the global intercept tend towards something very negative such that the intensity would tend toward 0 for these areas. For admissible regions, the specific intercept would more than offset this global intercept. I indeed initially tried with `-Inf `but `bru()` did not seem very happy about that.

> The proper way to handle known inadmissible regions is to specify a sampler polygon/s that only covers admissible regions; you'd need the intersection polygon of your current sampler boundary and the polygons for admissible regions.

I think at some point I tried that one before going for the current version. I changed because I thaught the `sampler` argument was intended to account for the case where you sampled only a fraction of the domain but points could occur everywhere and not only in the sampled areas (vs my case in which we sampled the whole domain and we know that we cannot observe points in some areas).  Thank you for clarifying.

In addition, one of the point that caused me troubles when I was using the `sampler` argument is that the population  raster is equal to 0 for inadmissible areas and therefore I ended up with messages such as `Non-finite (-Inf/Inf/NaN) entries detected in predictor derivatives for 'space'; treated as 0.0.`  during the evaluation step. This made inlabru crash and prompted me to come up with the version I presented in my initial post (I tried with the last devel version and the same problem happened). Should I keep `log(pop)=0` for inadmissible area while using the `sampler `argument to avoid that issue (I guess 'no')? Or maybe remove the offset and add the population as an additional linear component in the log intensity?



Best,
Olivier

**Reply by AlexandreBSimas** [reply]
Hi Olivier, as Finn mentioned, not only faster to build the mesh, but massively faster to build the metric graph object as well, along with other operations. You should try it again. 

As Finn mention, there will be a direct support with inlabru in the near future. 

We are also providing an alternative way to fit (that is turning out to be much faster than the finite element one) using the exact models, that we will also be providing in the near future as well.

Please, stay tuned, there will be lots of news regarding LGCP in metric graphs.

**Reply by finnlindgren** [reply, marked as answer]
Having a global intercept pushed to -Inf in the inadmissable region is a problem, as it then forces other components towards +Inf in the admissable region to compensate. You might have better luck splitting the intercept into one indicator for inside the admissible region, and a separate indicator for outside the admissible region. That might allow the second one to go to large negative values without interfering with the quantities of interest.

But you wouldn't need that -Inf component at all if you made the sampler the intersection of your current polygon an a polygon encapsulating the regions where population is > 0, as that would eliminate the need for "estimating" -Inf outside that region.  If you're worried that tyou would get non-zero estimates in that region, that's resolved by just setting the posterior samples to zero there; _you_ know that zero intensity is why you don't need to include it in the observation model, even though inlabru won't know about that.

**Reply by osupplisson** [reply]
Thank you very much to both of you for your helpful replies. I'll try both approaches (metricGraph/sampler with inlabru) to see what is the most feasible.

Best,
Olivier
